### DISCUSSION TITLE
[SPRING SECURITY] SecurityConfig 알아보기

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cs-collections/cs-collections-discussion#11] [Spring]
### 기존의 Security Config 클래스는

```java
public class WebSecurityConfig extends WebSecurityConfigurerAdapter 
```

_**처럼 WebSecurityConfigurerAdapter를 상속받아 메소드를 Override 하여 사용하였는데**_

<img width="951" alt="image" src="https://user-images.githubusercontent.com/65801290/207757960-de46f4bc-b711-41d7-97c6-ec29dc20360a.png">

_**보는 것과 같이 Deprecated 되었다. 
Spring Security 공식 홈피를 보게 되면**_

![image](https://user-images.githubusercontent.com/65801290/207758507-a96a8c61-21f9-4e59-a0f8-03508d6edb41.png)

_이제는 Bean 방식을 이용한 SecurityFilterChain 이용을 권장하고 있다._

_https://spring.io/blog/2022/02/21/spring-security-without-the-websecurityconfigureradapter_

_이제는 모든 설정을 빈 방식으로 채택한다는 말인데 예시로 인증을 무시할 경로들을 설정하는 것도 마찬가지로_

```java
    @Override
    public void configure(WebSecurity web) {
        web.ignoring().antMatchers("/ignore1", "/ignore2");
    }
```
에서

```java
    @Bean
    public WebSecurityCustomizer webSecurityCustomizer() {
        return (web) -> web.ignoring().antMatchers("/ignore1", "/ignore2");
    }
```
**_이런 식으로 바뀌게 된 것이다 이제부터 유의해서 설정을 해야한다._**

### _Security Filter Chain 사용법에 대해 알아보자._
```java
    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception { //(1)
   http
          .authorizeRequests() // (2)
            .antMatchers("/login", "/signup", "/user").permitAll() // 누구나 접근 허용
            .antMatchers("/admin/**").hasAnyRole("ADMIN") // admin으로 시작하는 모든 URL은 ADMIN 권한을 가진 사용자만 접근 허용.
            .antMatchers("/").hasRole("USER") // USER, ADMIN만 접근 가능
            .antMatchers("/admin").hasRole("ADMIN") // ADMIN만 접근 가능
            .anyRequest().authenticated() // 그 외 모든 요청에 대해 인증 필요
        .and() 
          .formLogin() // 로그인 페이지와 기타 로그인 처리 및 성공 실패 처리를 사용하겠다는 의미
            .loginPage("/login") // (3)
            .usernameParameter("id") // post 요청의 request id 값 설정
            .passwordParameter("pwd") // post 요청의 request pwd값과 같은 이름 설정
            .loginProcessingUrl("/login-process") // (4)
            .successHandler(loginSuccessHandler(null)) // 정상 인증 성공 후 별도의 처리가 필요 한 경우 핸들러를 생성하여 등록.
            .failureHandler(loginFailureHandler(null)) // 인증 실패 후 별도의 처리가 필요한경우 커스텀 핸들러를 생성하여 등록.
            .defaultSuccessUrl("/") // 로그인 성공 후 리다이렉트 주소
            .failureUrl("/login-fail") //인증이 실패 했을 경우 이동하는 페이지를 설정.
        .and()
          .logout() 
            .logoutSuccessUrl("/login") // 로그아웃 성공시 리다이렉트 주소
	    .invalidateHttpSession(true); // 세션 날리기

        http.headers().frameOptions().sameOrigin(); // (5)

        return http.build();
    }
```

(1) http 관련 인증 설정

(2) 특정 리소스의 접근 허용 또는 특정 권한을 가진 사용자만 접근을 가능하게 하는 설정.
> permitAll() : 누구나 접근이 가능
> hasRole() : 특정 권한이 있는 사람만 접근 가능
> authenticated() : 권한이 있으면 무조건 접근 가능
> anyRequest는 anyMatchers에서 설정하지 않은 나머지 경로를 의미함.

(3) 로그인 페이지 링크,  사용자가 따로 만든 로그인 페이지를 사용하려고 할때 설정. 따로 설정하지 않으면 디폴트 URL이 “/login”이기 때문에 “/login”로 호출하면 스프링이 제공하는 기본 로그인페이지가 호출된다.

(4) 로그인 FORM에서 아이디와 비번을 입력하고 확인을 클릭하면 “/login-process” 를 호출 하게 되었을 때 인증처리하는 필터가 호출되어 아이디 비번을 받아와 인증처리가 수행된다. UsernamePasswordAuthenticationFilter가 실행 되게 되는 것.

(5) X-Frame-option에 관련된 설정 
>sameOrigin : 같은 도메인일때만 iframe을 허용한다..?

향후 csrf등 추가 예정.

**Comment by JoisFe** [comment]
좋은 자료 감사합니다. 
궁금한점이 이전 AuthenticatioinManager, AuthenticationProvider 설정 방법이 configure메서드 오버라이딩으로 하였는데 Bean등록 방식으로 어떻게 하면 될까요?

**Reply by ahngilwoong** [reply]
![image](https://user-images.githubusercontent.com/65801290/207992120-fe498ef6-ca90-45e7-8683-f056d17fbf18.png)
 AuthenticatioinManager, AuthenticationProvider 설정 방법도 마찬가지로 위의 Bean등록 방식으로 설정하게 됩니다.

**Reply by JoisFe** [reply]
감사합니다.^^
